### PR TITLE
Translations for Electron main process

### PR DESCRIPTION
**Pull request by fragoulin**
Following this issue : https://github.com/reZach/i18next-electron-fs-backend/issues/3

Starting from your commit, the only missing stuff I noticed was :

- utf8 option for fs.readFile()
- call to JSON.parse() on data returned by fs.readFile

It's now working on my project with a single i18next config

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -266,7 +266,7 @@ class Backend {
             }
         } else {
             for (let i = 0; i < toWork.length; i++) {
-                fs.readFile(toWork[i].key, (error, data) => {
+                this.fs.readFile(toWork[i].key, "utf8", (error, data) => {
```
Comment (reZach) [inline src/index.js:269]: Why did you add `this` here?

File: src/index.js [hunk 2]
```diff
@@ -350,20 +350,31 @@ class Backend {
             loadPath,
             i18nextElectronBackend
         } = this.backendOptions;
-        let filename = this.services.interpolator.interpolate(loadPath, {
+        const filename = this.services.interpolator.interpolate(loadPath, {
```
Comment (reZach) [inline src/index.js:353]: Good suggestion here!

File: src/index.js [hunk 3]
```diff
@@ -350,20 +350,31 @@ class Backend {
...
             lng: language,
             ns: namespace
         });
 
+        // Handle initImmediate option
```
Comment (reZach) [inline src/index.js:358]: Can you explain a little bit more what this section does?

File: src/index.js [hunk 4]
```diff
@@ -130,6 +138,10 @@ class Backend {
             i18nextElectronBackend
         } = this.backendOptions;
 
+        if (typeof i18nextElectronBackend === "undefined"){
```
Comment (reZach) [inline src/index.js:141]: Instead of this, can you make an else for the `if` on line 122, and within this else put `this.setupIpcBindings();`?

Can you also put a comment above that line saying "Because we have the i18nextElectronBackend, we know we are in the renderer process and need to set up ipc bindings"

File: src/index.js (hunk 1 above)
Comment (fragoulin) [reply src/index.js:269]: I had an error during some tests regarding an unexisting fs at this line.
My bad I'll remove it.

File: src/index.js (hunk 3 above)
Comment (fragoulin) [reply src/index.js:358]: Default value for initImmediate is true. This means resources are loaded asynchronously.
If user decides to set it to false, resources are loaded synchronously.
More info here : https://www.i18next.com/overview/configuration-options

File: .babelrc [hunk 5]
```diff
@@ -1,4 +1,6 @@
 {
-    "presets": ["@babel/preset-env", "minify"],
+    "presets": ["@babel/preset-env"],
```
[the remark this answers is not in the document]
Comment (fragoulin) [reply .babelrc:2]: Yes

File: src/index.js [hunk 6]
```diff
@@ -277,17 +347,25 @@ class Backend {
...
-            if (error) return callback(error, false); // no retry
-            callback(null, data);
-        });
+        if (typeof i18nextElectronBackend !== "undefined") {
+            this.requestFileRead(filename, (error, data) => {
+                if (error) return callback(error, false); // no retry
+                callback(null, data);
+            });
+        } else {
+            this.fs.readFile(filename, (error, data) => {
```
[the remark this answers is not in the document]
Comment (fragoulin) [reply src/index.js]: Yes

File: src/index.js [hunk 7]
```diff
@@ -211,45 +223,103 @@ class Backend {
...
-        for (let i = 0; i < toWork.length; i++) {
-            var anonymous = function (error, data) {
-                if (error) {
-                    console.error(`${this.rendererLog} encountered error when trying to read file '${filename}' before writing missing translation ('${key}'/'${fallbackValue}') to file. Please resolve this error so missing translation values can be written to file. Error: '${error}'.`);
-                    return;
-                }
-
-                let keySeparator = !!this.i18nextOptions.keySeparator; // Do we have a key separator or not?
-                let writeKeys = [];
+        if (typeof i18nextElectronBackend !== "undefined") {
```
[the remark this answers is not in the document]
Comment (fragoulin) [reply src/index.js:226]: Yes

File: src/index.js [hunk 8]
```diff
@@ -211,45 +223,103 @@ class Backend {
...
+                    i18nextElectronBackend.send(writeFileRequest, {
+                        keys: writeKeys,
+                        filename: toWork[i].key,
+                        data
+                    });
+                }.bind(this);
+                this.requestFileRead(toWork[i].key, anonymous);
+            }
+        } else {
+            for (let i = 0; i < toWork.length; i++) {
```
[the remark this answers is not in the document]
Comment (fragoulin) [reply src/index.js:268]: Yes

File: src/index.js [hunk 9]
```diff
@@ -211,45 +223,103 @@ class Backend {
...
+                    // https://stackoverflow.com/a/51721295/1837080
+                    let separator = "/";
+                    const windowsSeparator = "\\";
+                    if (toWork[i].key.includes(windowsSeparator)) separator = windowsSeparator;
+                    let root = toWork[i].key.slice(0, toWork[i].key.lastIndexOf(separator));
+
+                    this.fs.mkdir(root, {
+                        recursive: true
+                    }, (error) => {
+                        fs.writeFile(toWork[i].key, JSON.stringify(data), (error) => {
```
[the remark this answers is not in the document]
Comment (fragoulin) [reply src/index.js]: UTF8 encoding is the default when it is not specified for fs.writeFile().
https://nodejs.org/api/fs.html#fs_fs_writefile_file_data_options_callback

File: src/index.js (hunk 4 above)
Comment (fragoulin) [reply src/index.js:141]: Not sure to understand what you want to achieve here.
If you need to refactor your code, I guess it's best to do it outside this PR.

File: src/index.js (hunk 3 above)
Comment (reZach) [reply src/index.js:360]: Great, thanks for explaining this to me, I was unsure where this came from. I see it must be necessary for loading translations in the main electron process now.

**Comment by reZach** [comment]
Thank you @fragoulin for your PR. I've posted some comments but I think once these are resolved I can merge these in and push a new version of this package live.

A question, you had said now your app only works with one config file with these changes, are you using [`secure-electron-template`](https://github.com/reZach/secure-electron-template) as your base template?

File: src/index.js (hunk 4 above)
Comment (reZach) [reply src/index.js:141]: I'd like to move the responsibility of the caller to determine whether or not to call `setupIpcBindings`, quite literally this:

```
// For cases we want to use this translator in the backend of Electron;
// ie menus and context menus, we need to be able to read/write translation
// files without using the window (it just doesn't exist!), so we dynamically
// import 'fs' here so that we are able to translate
if (typeof this.backendOptions.i18nextElectronBackend === "undefined") {
    this.fs = require("fs");
} else {

    // Because we have the i18nextElectronBackend, we know we are in the renderer process and need to set up ipc bindings
    this.setupIpcBindings();
}
```

instead of the line you had added here:
```
if (typeof i18nextElectronBackend === "undefined") {
    return;
}
```

If you want to just copy all the code I pasted above, that'd be great, otherwise I can simply make the change myself.

Comment (fragoulin) [reply src/index.js:141]: I think it's better if you do it yourself, because I didn't add this line myself :
```
if (typeof i18nextElectronBackend === "undefined") {
    return;
}
```
I simply followed your direction and started from your commit https://github.com/reZach/i18next-electron-fs-backend/commit/1af4d358a260873b92b1af084dc15668660736f4

You did all the hard work and I simply had to fix a couple of little issues to make your changes works. I'm not confident to do some refactoring on your code and you have a better view than I have on your code.

**Comment by fragoulin** [comment]
> 
> 
> Thank you @fragoulin for your PR. I've posted some comments but I think once these are resolved I can merge these in and push a new version of this package live.
> 
> A question, you had said now your app only works with one config file with these changes, are you using [`secure-electron-template`](https://github.com/reZach/secure-electron-template) as your base template?

I think I've answered all your comments.
I don't use [`secure-electron-template`](https://github.com/reZach/secure-electron-template) as my base template because I use electron-forge, but I took some parts here and there and I have the same security configuration. Sorry but I can't publish my code yet.
If you have trouble integrating those changes to your template, I could have a look.

File: src/index.js (hunk 4 above)
Comment (reZach) [reply src/index.js:141]: _Oh_! I forgot I even wrote this line - thanks for being patient with me here. As you can see it's obviously been a lot of time since I've looked at this.

**Comment by reZach** [comment]
> 
> 
> > Thank you @fragoulin for your PR. I've posted some comments but I think once these are resolved I can merge these in and push a new version of this package live.
> > A question, you had said now your app only works with one config file with these changes, are you using [`secure-electron-template`](https://github.com/reZach/secure-electron-template) as your base template?
> 
> I think I've answered all your comments.
> I don't use [`secure-electron-template`](https://github.com/reZach/secure-electron-template) as my base template because I use electron-forge, but I took some parts here and there and I have the same security configuration. Sorry but I can't publish my code yet.
> If you have trouble integrating those changes to your template, I could have a look.

Great thanks, there's just _one_ comment on your PR where I believe you added `this` and if you can revert your changes to **.babelrc** that'd be great (I can go back and add comments later in the code).

Hm, I was hoping you used the template, I had trouble with my own, maybe I'll give it a go once your PR is merged.

File: src/index.js (hunk 1 above)
Comment (fragoulin) [reply src/index.js]: Done in commit https://github.com/reZach/i18next-electron-fs-backend/pull/4/commits/4728f30698228602ad05164e84b4832c99b95e73

File: .babelrc (hunk 5 above)
[the remark this answers is not in the document]
Comment (fragoulin) [reply .babelrc]: Done in commit https://github.com/reZach/i18next-electron-fs-backend/pull/4/commits/4728f30698228602ad05164e84b4832c99b95e73
